### ISSUE TITLE
Enhance randomized process tests with start process instance anywhere ability

### ISSUE THREAD

**Issue by korthout** [camunda/zeebe#9407]
To comprehensively test the start process instance anywhere feature, we'll need to enhance the randomized process tests with the ability to start the process instance anywhere.

Blocked by #9390 #9391

## Out of scope
- starting the process instance inside a multi-instance
- starting the process instance inside a call activity

**Comment by pihme** [comment]
### Summary of discussions

tldr: it's probably possible but maybe not worth it.

#### Technical background
* The current random process generation does not randomly generate all possible workflows. All workflows follow a certain well-defined structure in which each branching element also has a corresponding joining element. This structure represents only a subset of all possible processes. And it makes it possible for us to generate an execution plan for each process that has this structure.
* This structure also becomes tricky when trying to find which elements to start from: You cannot just pick one element in one parallel gateway branch; you need to pick elements for all branches
* Technically it probably could be done if we pick the start positions in the recursive process that generates the execution path. Then for each such pick, we would discard any steps which would have been generated before it, and only keep the steps
* This also works nice with the way start variables are flushed to the start of the execution path. These variables are used to choose which branch to take in an exclusive gateway, and which timeout to set for timer events. The variables could be set by the command to start the process instance from anywhere instead of the usual start event

#### Reasons against implementing it
* This is already a complicated piece of code and this would make it even more complicated and difficult to understand
* The limitations in terms of structures which are generated reduce the usefulness of the random tests. Most real world bugs we found, where a) in processes which don't follow this neat structure, and b) in combination with non-trivial variable expressions and nested variable scopes. So the bottom line is, we would need additional handcrafted test cases anyway to test this feature
* It is unclear, but also very unlikely, that the approach sketched above will also work with future extensions to start process from anywhere, in particular different start instructions and more complicated instructions to set variables in certain scopes.

**Comment by remcowesterhoud** [comment]
Just wanted copy the solution I described in my PR here as well for completeness sake:

**Concept**
When we generate a random process all we do is create a graph of BlockBuilders. By looking at the BlockBuilders you almost see exactly what the process looks like. In the end almost all elements in a process resolves to being a single BlockBuilder.

When we want to generate an execution path there really is only 3 pieces of information we need:
1. The block we want to start at
2. Whether or not this block is part of the "branch" of BlockBuilders we are currently generating the path for
3. Whether or not this block comes after the start block in the process

Knowing this we can have a look at a simple process:
<img width="1065" alt="image" src="https://user-images.githubusercontent.com/5787702/176442280-47cabcee-f967-4ccc-90f9-61f3b82f6b3e.png">

Imagine we want to start at **task 2**. Just looking at the process we know that we want to generate execution steps for task 2 and task 4. After this the process is completed. Programmatically we can do this using the 3 pieces information we have available.

We know the start block, this is `BlockBuilder 4`. Each BlockBuilder is aware of the itself, and BlockBuilder that are nested in here. This means we can check whether or not the start block is part of the current "branch" we want to create an execution path for. In this case:

- BB1 has BB4 as a nested block
- BB2 does not have BB4 as a nested block
- BB3 has BB4 as a nested block
- BB4 is our start block
- BB5 does not have BB4 as a nested block
- BB6 does not have BB4 as a nested block

With this logic we can step-by-step create an execution path starting from the highest-level block builder (BB1):
1. Find a path for BB1. Our starting block is nested within BB1, so we should generate an execution path for this block. Since BB1 has no relevant elements (we can ignore start and end events here) it doesn't have to generate any steps itself. Since it does contain other block builders it will loop over these in order and have them all generate their execution path if required. This means that the next block generating an execution path will be BB2.
2. Find a path for BB2. BB2 != BB4 and does not contain BB4 as a nested block. It also does not appear in the process after BB4. We can ignore this block and don't need to generate an execution path for it. We continue BB1's loop and go to BB3.
3. Find a path for BB3. The starting block is nested within BB3, so we should generate an execution path for this block.Since BB3 != BB4 we can ignore the forking gateway, as it occurs before our starting block. The ExclusiveGatewayBlockBuilder contains 2 branches. A branch is nothing more than another block builder. We can check for each branch whether or not it has BB4 nested within. If it does we will generate an execution path for this branch. If it doesn't we can fully ignore this branch. This means we will now continue to find a path for BB4. BB5 will never search for a path.
4. Find a path for BB4. BB4 is our starting block! We can detect this and finally start generating execution steps. We will also set a flag to true, indicating that any remaining blocks will be behind BB4 and should also generate execution steps. After this BB1's loop will continue and we will go to BB6.
5. Find a path for BB6. BB6 != BB4, nor is BB4 nested within. However, since we turned the flag to true we know that we have already passed BB4. Therefore BB6 knows that it should generate execution steps.

In the end we will end up with a list containing all the execution steps that need to be taken. Since only BB4 and BB6 have added any steps to this list we can safely start at BB4 and the process will complete fully. Generation of variables and messing with timers is not necessary and behave the same way as it does when starting from the start event.

**Parallel gateway**
Parallel gateways are an exception to the rule. If we pretend that the exclusive gateway from the simple process is now a parallel gateway we need some different behaviour. We can nog longer ignore the branch of BB5, because if we do and we start at BB4 the process will get stuck at the joining gateway. To circumvent this the ParallelGatewayBlockBuilder does a little trick. It will first generate the path for branch containing the starting block. This will set the flag that we have passed the starting block to true. Afterwards it will generate the path for all other branches. Because the flag is set to true these branches will add all the steps that they can. Finally it will also add the first block builder of each of those remaining branches to the ExecutionPathContext. When sending the start command we can access these blocks and add the element id of the first element in this block to the start instructions. As a result we will always fully execute those branches.
